synthesis/codon/codon.go

```diff
@@ -163,45 +155,149 @@ func Optimize(aminoAcids string, codonTable Table, randomState ...int) (string,
 	}
 
 	var codons strings.Builder
-	codonChooser, err := codonTable.Chooser()
-	if err != nil {
-		return "", err
-	}
+	codonChooser := table.Choosers
 
 	for _, aminoAcid := range aminoAcids {
 		chooser, ok := codonChooser[string(aminoAcid)]
 		if !ok {
 			return "", invalidAminoAcidError{aminoAcid}
 		}
+
 		codons.WriteString(chooser.Pick().(string))
 	}
+
 	return codons.String(), nil
 }
 
-// OptimizeTable weights each codon in a codon table according to input string codon frequency.
-// This function actually mutates the codonTable struct itself.
-func (table codonTable) OptimizeTable(sequence string) Table {
+// UpdateWeights will update the translation table's codon pickers with the given amino acid codon weights
+func (table *TranslationTable) UpdateWeights(aminoAcids []AminoAcid) error {
+	// regenerate a map of codons -> amino acid
+
+	var updatedTranslationMap = make(map[string]string)
+	for _, aminoAcid := range table.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			updatedTranslationMap[codon.Triplet] = aminoAcid.Letter
+		}
+	}
+
+	table.TranslationMap = updatedTranslationMap
+
+	// Update Chooser
+	updatedChoosers, err := newAminoAcidChoosers(table.AminoAcids)
+	if err != nil {
+		return err
+	}
+
+	table.Choosers = updatedChoosers
+	table.AminoAcids = aminoAcids
+
+	return nil
+}
+
+// UpdateWeightsWithSequence will look at the coding regions in the given genbank data, and use those to generate new
+// weights for the codons in the translation table. The next time a sequence is optimised, it will use those updated
+// weights.
+//
+// This can be used to, for example, figure out which DNA sequence is needed to give the best yield of protein when
+// trying to express a protein across different species
+func (table *TranslationTable) UpdateWeightsWithSequence(data genbank.Genbank) error {
+	codingRegions, err := extractCodingRegion(data)
+	if err != nil {
+		return err
```

Comment:
not covered by tests.

Reply:
`extractCodingRegion` would only error due to a `feature.GetSequence()` call, which never errors as per [this comment](https://github.com/TimothyStiles/poly/pull/350/files#r1386924848)